### ISSUE TITLE
Error while grading test

### ISSUE THREAD

**Issue by Alex-Jordan** [openwebwork/webwork2#2148]
In 2.18, I made a test. May not be relevant, but it uses a proctoring password. The test has 33 exercises, 6 per page. There are 10 total tikz images (just mentioning because they take time to load). There are three essay answer fields (is that allowed in a test?)

I made a fake student account and took the test. After answering everything, I clicked to Grade. I clicked to confirm. Then the page hung for a long time. I figured that must be normal, since it must be rendering all the problems including those tikz images, just for the answer checking. But then an error screen came (contents pasted below at the end). 

Still as the student, I returned to the test. I was prompted to again enter the password to return to the test. That seemed wrong because I had submitted it and it is configured to only allow one graded submission. But the error message might indicate some process wasn't concluded. Anyway I entered the password and it hung for a long time. Too long. I got impatient and clicked to go back to the test's landing page. This time I saw this:

<img width="935" alt="Screen Shot 2023-07-24 at 1 52 42 AM" src="https://github.com/openwebwork/webwork2/assets/4732672/64d3fa67-28f8-416e-b0d6-870777b994e1">

It is "completed" but "not yet submitted". Something is off there.

Lastly, I thought I was answering it all correctly, except of course for the essay questions that would need manual grading. So I was surprised to see my score is 138/200. Looking more closely, starting at number 25, I have all 0s. So it seems that scoring was interrupted starting there. That problem uses a tikz image. I clicked to view it and it hung, eventually giving me the error message `malformed JSON string, neither tag, array, object, number, string or atom, at character offset 0 (before "(end of string)") at /usr/local/share/perl5/Mojo/JSON.pm line 31. ` But I am able to click to edit the problem and it loads fine in the editor. Even with the same seed as was used in the test.

Here is the error message:

```
WeBWorK error
An error occured while processing your request.

For help, please send mail to this site's webmaster webwork-group@pcc.edu, including all of the following information as well as what what you were doing when the error occured.

Error record identifier
39f5457c-1988-5a39-a881-60d1dd21acfd::1285033a-29ff-11ee-8630-9afc043bea0e

Error messages
Can't use an undefined value as a HASH reference at /opt/webwork/webwork2/lib/WeBWorK/Utils/ProblemProcessing.pm line 328.

Context
323: 	my @answerTypes;
324: 	my %answers_to_store;
325: 	my @past_answers_order;
326: 	my @last_answer_order;
327: 
328: 	my %pg_answers_hash = %{ $pg->{PG_ANSWERS_HASH} };
329: 	foreach my $ans_id (@{ $pg->{flags}{ANSWER_ENTRY_ORDER} // [] }) {
330: 		$scores2 .= ($pg_answers_hash{$ans_id}{rh_ans}{score} // 0) >= 1 ? "1" : "0";
331: 		push @answerTypes, $pg_answers_hash{$ans_id}{rh_ans}{type} // '';
332: 		foreach my $response_id (@{ $pg_answers_hash{$ans_id}{response_obj}{response_order} }) {
333: 			$answers_to_store{$response_id} = $problem->{formFields}{$response_id};
Call stack
Request information
```

**Comment by somiaj** [comment]
Yes, there were multiple problems with TiKz images in the test (I also had a few essay questions).

My test only gave one problem per page (so that is a bit different than @Alex-Jordan issue) and out of 10 students who took the test, only one had this issue (maybe they submitted on a problem that different students didn't). I also don't recall any error for that student, so the affect (if they are related) might have a different cause.

If the underlying issue isn't related, this means there might be multiple ways in which the problems can be counted as attempted but the test completion time isn't saved.

**Comment by Alex-Jordan** [comment]
Here is the one problem that (still) won't load when I try to view. Note that I can view it from the editor. And I could see it when I was taking the test. But if I go to the Set Details page now and click the eyeball "Open in a New Window" icon, it hangs and eventually gives the "malformed JSON" error posted above. And coincidentally or not, this is where all scores start being 0.

[testWithTikzProblem.txt](https://github.com/openwebwork/webwork2/files/12147661/testWithTikzProblem.txt)

**Comment by Alex-Jordan** [comment]
When I trigger the "malformed JSON" error, there is nothing added to webwork2.log.

I do see the original message in the log though, from late last night.

**Comment by drgrice1** [comment]
I am rather certain that what is happening is that there is a LaTeX error that is occurring (it seems to be dependent on the seed for the problem you posted).  When that LaTeX error happens the LaTeXImage.pm code stops and waits for input.  Then the request times out, and hence processing of the request ceases.  Thus you get 0's for the remaining problems.

**Comment by Alex-Jordan** [comment]
The seed that is used in the scenarios where I get "malformed JSON" is 1798.

**Comment by drgrice1** [comment]
I am not getting an error with that seed for the problem.  However, I created a test that contained a problem that had TikZ code that was guaranteed to generate an error (it was missing a semicolon at the end of a line in the TikZ code).  When I graded the test the request hung and eventually timed out.  The error that was reported was the one that you posted above.  Then I added `--interaction=nonstopmode` to the LaTeXImage.pm code (and to be sure reset everything -- reassigned the quiz to my user and deleted temporary image files).  When I grade the test with that, the grading completes fine, and the errors from LaTeXImage.pm are displayed on the screen.

**Comment by Alex-Jordan** [comment]
If anyone would like to see if you can replicate what happens on my system, here is the test and its problem files. You may need to change the open date before a fake student can take it.

[mth095-alex_jordan-202303-assessment.tgz](https://github.com/openwebwork/pg/files/12157824/mth095-alex_jordan-202303-assessment.tgz)

**Comment by Alex-Jordan** [comment]
So it seems that grading this test is more than is needed for me to hit on the issue.

With a student account, I can start this test. I go straight to page 5 (with numbers 25–30). I type the answers to number 25. I go to page 6. I try to go back to page 5, and it hangs until it gives the error message.

Maybe there is something bad about the code I have for number 25. I haven't seen it yet, but I'm still looking. Nevertheless, a bug in a problem shouldn't cause this behavior.

**Comment by Alex-Jordan** [comment]
Definitely an issue with this problem file. In the editor I can type in answers and click to check answers, and it hangs.

**Comment by Alex-Jordan** [comment]
I whittled it down to where the problem calls the `reduce` method on the Formula in FiniteSolutionSets context. Removing that stops this issue. So this is some bug with contextFiniteSolutionSets.pl.

**Comment by drgrice1** [comment]
I am closing this again.  This was fixed by #2154.  Well, the webwork2 issue caused by using the system Safe at least.  Not the PG issue with the `contextFiniteSolutionSets.pl` macro.  An issue can be opened for that on the PG repository if so desired.
